### PR TITLE
feat: Add `round_sig_figs` expression for rounding to significant figures

### LINKED ISSUE #11968: Rounding to significant figures
### Description

Recently I have been working on a project where we needed to round to significant figures and realised there is no way to do this in pandas, numpy or polars. I found [this](https://stackoverflow.com/questions/18915378/rounding-to-significant-figures-in-numpy/59888924#59888924) as the best solution.

```
def signif(x, p):
    x = np.asarray(x)
    x_positive = np.where(np.isfinite(x) & (x != 0), np.abs(x), 10**(p-1))
    mags = 10 ** (p - 1 - np.floor(np.log10(x_positive)))
    return np.round(x * mags) / mags
```

However it would be nice to have it as part of a package. Would it be possible to include this? 

I have implemented this already here #11959 which is about 20% faster than the numpy one here already. However I can imagine there is a better way than using the zip_with maybe? 

Thank you!

### PR DESCRIPTION
**Pull request by owrior**
Recently I have been working on a project where we needed to round to significant figures and realised there is no way to do this in pandas, numpy or polars. I found [this](https://stackoverflow.com/questions/18915378/rounding-to-significant-figures-in-numpy/59888924#59888924) as the best solution.

```
def signif(x, p):
    x = np.asarray(x)
    x_positive = np.where(np.isfinite(x) & (x != 0), np.abs(x), 10**(p-1))
    mags = 10 ** (p - 1 - np.floor(np.log10(x_positive)))
    return np.round(x * mags) / mags
```

However it would be nice to have it as part of a package. Would it be possible to include this? (The implementation here is about 20% faster than the numpy one here already but I imagine there is a better way of implementing it) (Update now about 45% faster).

Thank you!

Closes #11968

### CODE REVIEW

File: crates/polars-ops/src/series/ops/round.rs [hunk 1]
```diff
@@ -37,6 +37,55 @@ pub trait RoundSeries: SeriesSealed {
...
+        if let Ok(ca) = s.f32() {
+            let s = ca
+                .zip_with(
+                    &ca.equal(0),
+                    &ca.apply_values(|val| round_sf(val as f64, significant_figures) as f32),
+                )?
+                .into_series();
+            return Ok(s);
+        }
+        if let Ok(ca) = s.i32() {
```
Comment (owrior) [inline crates/polars-ops/src/series/ops/round.rs:68]: Is this the right way to allow for it working with integers?

**Comment by stinodego** [comment]
Hi @owrior, thanks for the PR. Could you please make an issue so we can discuss whether we want to include this, and what the right API for it would be? Thanks!

From a quick initial look, the boilerplate here looks quite good 👌

**Comment by owrior** [comment]
Refactored to round_sig_figs @orlp!

File: crates/polars-ops/src/series/ops/round.rs [hunk 2]
```diff
@@ -37,6 +37,43 @@ pub trait RoundSeries: SeriesSealed {
         polars_bail!(opq = round, s.dtype());
     }
 
+    fn round_sig_figs(&self, significant_figures: u32) -> PolarsResult<Series> {
+        let s = self.as_series();
+        let significant_figures = if significant_figures == 0 {
```
Comment (owrior) [inline crates/polars-ops/src/series/ops/round.rs:42]: Should I raise an exception of 0 is specified? How should this be done? Is it polars bail?

Comment (owrior) [reply crates/polars-ops/src/series/ops/round.rs]: I have done this.

File: crates/polars-ops/src/series/ops/round.rs (hunk 1 above)
Comment (orlp) [reply crates/polars-ops/src/series/ops/round.rs:62]: Could you use the `with_match_physical_numeric_polars_type` macro to cover all types?

File: crates/polars-ops/src/series/ops/round.rs [hunk 3]
```diff
@@ -70,6 +107,14 @@ pub trait RoundSeries: SeriesSealed {
 
 impl RoundSeries for Series {}
 
+fn round_sig_figs(value: f64, digits: u32) -> f64 {
+    if value == 0.0 {
+        return value;
+    }
+    let magnitiude = 10.0.pow(digits as f64 - 1.0 - ((value.abs()).log10().floor()));
```
Comment (orlp) [inline crates/polars-ops/src/series/ops/round.rs:114]: Could you use `f64::powi`? Also it's spelled `magnitude`.

File: py-polars/polars/expr/expr.py [hunk 4]
```diff
@@ -1765,6 +1765,34 @@ def round(self, decimals: int = 0) -> Self:
         """
         return self._from_pyexpr(self._pyexpr.round(decimals))
 
+    def round_sig_figs(self, digits: int = 1) -> Self:
```
Comment (orlp) [inline py-polars/polars/expr/expr.py:1768]: Why the default argument of 1? I'd personally not set a default at all.

File: py-polars/scripts/tes.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import time
```
Comment (orlp) [inline py-polars/scripts/tes.py:1]: Did you mean to commit this file to the repository? I don't think we want this.

File: crates/polars-ops/src/series/ops/round.rs [hunk 6]
```diff
@@ -37,6 +37,43 @@ pub trait RoundSeries: SeriesSealed {
         polars_bail!(opq = round, s.dtype());
     }
 
+    fn round_sig_figs(&self, digits: u32) -> PolarsResult<Series> {
+        let s = self.as_series();
+        if digits < 1 {
+            polars_bail!(
+                InvalidOperation: "Digits must be a positive integer."
```
Comment (orlp) [inline crates/polars-ops/src/series/ops/round.rs:44]: Could you change this to `"digits must be an integer >= 1"`?

**Review by orlp — CHANGES_REQUESTED** [review]

File: crates/polars-ops/src/series/ops/round.rs (hunk 3 above)
Comment (owrior) [reply crates/polars-ops/src/series/ops/round.rs]: i'm struggling to use the with_match_physical_numeric_polars_type macro. I cannot use $T to convert primitives so am not sure what the structure should look like to accomodate this.

```
        with_match_physical_numeric_polars_type!(s.dtype(), |$T| {
            let ca: &ChunkedArray<$T> = s.as_ref().as_ref().as_ref();
            if $T.is_float() {
                let s = ca.apply_values(|val| {
                    if val == 0.0 {
                        return 0.0;
                    }
                    let magnitude = 10.0_f64.powi(digits - 1 - ((value.abs()).log10().floor() as i32));
                    (value * magnitude).round() / magnitude
                }).into_series();
            } else {
                let s = ca.apply_values(|val| {
                    if val == 0 {
                        return 0;
                    }
                    let magnitude = 10.powi(digits - 1 - ((value.abs()).log10().floor() as i32));
                    (value * magnitude).round() / magnitude
                }).into_series();
            }

            return Ok(s);
        });
```

File: crates/polars-ops/src/series/ops/round.rs [hunk 7]
```diff
@@ -37,6 +38,28 @@ pub trait RoundSeries: SeriesSealed {
...
+            let ca = s.f64()?;
+            let s = ca.apply_values(|value| {
+                if value == 0.0 {
+                    return 0.0;
+                }
+                let magnitude = 10.0_f64.powi(digits - 1 - ((value.abs()).log10().floor() as i32));
+                (value * magnitude).round() / magnitude
+            }).into_series();
+            return Ok(s);
+        });
```
Comment (owrior) [inline crates/polars-ops/src/series/ops/round.rs:60]: Hi @orlp sorry its been a bit messy! I have gotten this to work with the type matching here. Let me know 1. if this is ok - not sure if the explicit casting to float64 in the middle is acceptable and 2. I am not sure how to convert back to the original datatype using $T - can you advise? Thank you.

Comment (owrior) [reply crates/polars-ops/src/series/ops/round.rs:60]: Also this is slower than the original implementation I had - still about 10% faster than the original numpy solution so not a big problem.

File: crates/polars-ops/src/series/ops/round.rs (hunk 3 above)
Comment (owrior) [reply crates/polars-ops/src/series/ops/round.rs]: Think this is done.

**Review by orlp — APPROVED** [review]

**Review by stinodego — CHANGES_REQUESTED** [review]
I will do a small update to the docstrings, will update and merge after.

**Review by stinodego — APPROVED** [review]

**Pull request merged**